### PR TITLE
feat: [CDS-43771]: Changes related to template reconciliation in custom deployment

### PR DESCRIPTION
**Pull request by sourabh-awashti**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/38034)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_NA_** 
 


            
 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/38034)

**Comment by sourabh-awashti** [comment]
trigger messagecheck

**Review by RishabhGupta34 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38034)* status: 7 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/customDeployment/EventListener/CustomDeploymentEntityCRUDEventHandler.java` line 46 at r3](https://reviewable.io/reviews/harness/harness-core/38034#-ND2ub_w6qEwJcfMXgWR:-ND2ub_w6qEwJcfMXgWS:b98xa2n) ([raw file](https://github.com/harness/harness-core/blob/8b4f53254e732e2d740e7a942bffdbe4731e87c9/125-cd-nextgen/src/main/java/io/harness/cdng/customDeployment/EventListener/CustomDeploymentEntityCRUDEventHandler.java#L46)):*
> ```Java
>         Scope.builder().accountIdentifier(accountRef).orgIdentifier(orgRef).projectIdentifier(projectRef).build();
>     String entityFQN = getFullyQualifiedIdentifier(accountRef, orgRef, projectRef, identifier) + "/";
>     if (!versionLabel.isEmpty()) {
> ```

versionLabel can be null as well, so this will give NPE

File: 125-cd-nextgen/src/main/java/io/harness/cdng/customDeployment/EventListener/CustomDeploymentEntityCRUDEventHandler.java [hunk 1]
```diff
@@ -42,8 +42,10 @@ public boolean updateInfraAsObsolete(
       String accountRef, String orgRef, String projectRef, String identifier, String versionLabel) {
     Scope scope =
         Scope.builder().accountIdentifier(accountRef).orgIdentifier(orgRef).projectIdentifier(projectRef).build();
-    String entityFQN = getFullyQualifiedIdentifier(accountRef, orgRef, projectRef, identifier);
-    entityFQN = entityFQN + "/" + versionLabel + "/";
+    String entityFQN = getFullyQualifiedIdentifier(accountRef, orgRef, projectRef, identifier) + "/";
+    if (versionLabel == null) {
```
Comment (RishabhGupta34) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/customDeployment/EventListener/CustomDeploymentEntityCRUDEventHandler.java:46]: nit. not equal to right?

**Review by RishabhGupta34 — APPROVED** [review]
:lgtm:

> Reviewed 4 of 7 files at r2, 2 of 2 files at r3, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38034#-:-ND3YCoHEM0Bf700Yoqw:bnfp4nl)* status: :shipit: complete! ✓

**Pull request merged**